Retry recycle or delete operation on failure.

jsafrane opened the pull request:
Recycle controller tries to recycle or delete a PV several times.
It stores count of failed attempts and timestamp of the last attempt in
~~annotations of the PV~~ internal map.

By default, the controller tries to recycle/delete a PV 3 times in
10 minutes interval. These values are configurable by
kube-controller-manager --pvrecycler-retry-count=X --pvclaimbinder-sync-period=Y
arguments.

cc: @kubernetes/rh-storage

derekwaynecarr commented:
@markturansky - can you pick this up?

markturansky commented:
race condition.  i assigned myself just as you posted.

yes, i will review this.

markturansky commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -81,6 +81,7 @@ type CMServer struct {
 	ResourceQuotaSyncPeriod           time.Duration
 	NamespaceSyncPeriod               time.Duration
 	PVClaimBinderSyncPeriod           time.Duration
+	PVRecycleRetryCount               int
```
I think this is better in the VolumeConfigFlags struct.

jsafrane replied:
Ok, moved.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -94,16 +116,61 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
+	}
+
+	// retry-count is fine, check the timestamp
+	lastFailureStr, ok := pv.Annotations[pvRecycleLastFailureAnnotationKey]
```
what are the benefits/consequences of using annotations to store this versus a local map?

A local map containing counts would remove annotation keys, remove PV updates, etc. but it wouldn't guarantee 3 retries.  A restarted controller manager would necessarily have an empty map and the PV could go 3 more times, even if it attempted recycling twice before the controller manager was restarted.

I don't think the guarantee of X times is important, especially if it makes the code smaller/easier for this feature.

jsafrane replied:
Local cache is probably easier to handle and does not generate 'pv updated' events to other controllers, which is much better. I'm adding a patch with local cache of failed volumes (to be squashed into the final patch eventually).

On the other side, with the local cache the state is hidden inside the controller and any debugging/troubleshooting is much harder (as I found out right now). 
Annotations in `kubectl get pv -o yaml` were quite useful.

jsafrane commented:
@k8s-bot unit test this

markturansky commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -85,6 +85,7 @@ type VolumeConfigFlags struct {
 	PersistentVolumeRecyclerMinimumTimeoutHostPath      int
 	PersistentVolumeRecyclerIncrementTimeoutHostPath    int
 	EnableHostPathProvisioning                          bool
+	PVRecycleRetryCount                                 int
```
For consistency, I think this should be `PersistentVolumeRecyclerRetryCount`, which would match the other variables in this struct.

jsafrane replied:
Ok

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -36,25 +36,60 @@ import (
+	// triggers new retry, which is likely to fail again.
+	pvRecycleLastFailureAnnotationKey = "volume.experimental.kubernetes.io/last-failure"
+)
+
```
These are no longer used/needed if you've got a local map storing state.

jsafrane replied:
ok

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -94,16 +136,56 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
-		if latest.Status.Phase != api.VolumeReleased {
-			return fmt.Errorf("PersistentVolume[%s] phase is %s, expected %s.  Skipping.", pv.Name, latest.Status.Phase, api.VolumeReleased)
+
+		if err = recycler.shouldRecycle(latest); err != nil {
```
Why is this here as well as on line 179 above?  This check is gated by the 1st `shouldRecycle` just a few lines earlier.

jsafrane replied:
Well... why it was there in the original code? I did not change its logic.

markturansky replied:
`shouldRecycle` is a new helper function you added to support retries.  Current `reclaimVolume` func does not contain it.

jsafrane replied:
If you look at `reclaimVolume()` before my patch, you see that the `pv.Status.Phase == api.VolumeReleased` is checked twice too. Now that the check is moved to `shouldRecycle` and extended.

markturansky replied:
You're right.   This line makes no sense, as it will never be true:

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go#L105

Unless I'm missing something, we don't need a 2nd check.  The 1st check can be broken in 2:

```
if phase != released, return error
if claimRef == nil, return error
...
the rest of the code...
```

jsafrane replied:
There is `pv = GetPersistentVolume(pv.Name)`  before the second checks, so I just guess you wanted double-check the PV did not change. Some other controller might modify the PV (it's quite busy object these days :-).

jsafrane replied:
I removed this double check in the last patch, I'll reload `pv` from the server and check it only once.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -123,10 +205,32 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 			glog.Errorf(errMsg)
 			return fmt.Errorf(errMsg)
 		}
+	} else {
```
I think you can drop `else

jsafrane replied:
logging in the `else` branch has proven to be valuable during testing, I'd like to leave as it is.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -151,10 +255,14 @@ func (recycler *PersistentVolumeRecycler) handleRecycle(pv *api.PersistentVolume
 			nextPhase = api.VolumeFailed
 		} else {
 			glog.V(5).Infof("PersistentVolume[%s] successfully recycled\n", pv.Name)
+			// The volume has been recycled. Remove any retry annotations to make
```
comment needs updating.  no more annotations for state.

jsafrane replied:
Good catch!

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -123,10 +205,32 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
+}
+
+func (recycler *PersistentVolumeRecycler) removeFailedVolume(pv *api.PersistentVolume) {
+	delete(recycler.failedVolumes, pv.Name)
```
Do we need `removeFailedVolume`?  This is a one-liner and so is the `deleteVolume` func.  Maybe just `deleteVolume`?

jsafrane replied:
Good point, `deleteVolume` removed

jsafrane commented:
@k8s-bot test this please

jsafrane commented:
@k8s-bot test this please

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -123,10 +188,28 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 			return fmt.Errorf(errMsg)
 		}
+		return nil
+	} else {
```
By drop `else`, I mean you can remove the else clause but keep the glog statement.  This can be flattened by un-indenting and then return nil.

jsafrane replied:
Oh, ok, that makes sense.

jsafrane commented:
rebased

jsafrane commented:
@k8s-bot unit test this

markturansky commented:
All green, Jan.  I'll do another detailed review today, now that the merge queue is unblocked.

jsafrane commented:
I forgot to push the 2nd patch when rebasing, pushed now.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -89,6 +107,13 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
+				if !ok {
+					glog.Errorf("Error casting object to PersistentVolume: %v", obj)
+				}
+				recycler.removeFailedVolume(pv)
```
Would need to return if !ok.   This function would fail if pv == DeletedFinalStateUnknown.

jsafrane replied:
Fixed

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -96,16 +121,57 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 	return recycler, nil
 }
 
+func (recycler *PersistentVolumeRecycler) shouldRecycle(pv *api.PersistentVolume) error {
```
please add a comment for what the func does.

jsafrane replied:
Added

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -96,16 +121,57 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
+	}
+	if pv.Status.Phase == api.VolumeReleased {
+		// This is the first attempt to recycle, allow.
+		return nil
```
I think this is confusing now.  I understand why you've done this, because the recycler will mark the PV failed if recycling fails and you check for that status below.   But that now means 2 different states represent a need to recycle:  status=Released || status=Failed && !exceedFailedCount.  

`kubectl get pv` would show PVs that failed, but only some are truly Failed.  

We should use Released with the count you're keeping.  When the count is exceeded, the PV is truly failed.

jsafrane replied:
I rewrote the patch to keep the volume `Released` and mark it as `Failed` only when all attempts to recycle/delete volume fail.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
+	lastAttempt time.Time
 }
 
 // PersistentVolumeRecycler creates a new PersistentVolumeRecycler
```
this is my typo in the original, but since you've got this PR open, would you fix the comment to start with the correct func name?

jsafrane replied:
Fixed

jsafrane commented:
Updated with @markturansky's comments and rebased.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
-// PersistentVolumeRecycler creates a new PersistentVolumeRecycler
-func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Duration, plugins []volume.VolumePlugin, cloud cloudprovider.Interface) (*PersistentVolumeRecycler, error) {
+// releasedVolumeStatus holds state of failed delete/recycle operation on a
+// volume. The controller re-tries the opration several times and it stores
```
typo: opration

jsafrane replied:
fixed

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
+	if status.retryCount > recycler.maximumRetryCount {
+		// This was the last attempt. Remove any internal state and mark the
+		// volume as Failed.
+		glog.V(3).Infof("PersistentVolume[%s] failed %dth time - marking Failed", pv.Name, status.retryCount)
```
"failed %d times" otherwise the log would read 2th or 3th.

swagiaal replied:
This has not been addressed it looks like

jsafrane replied:
Fixed

markturansky commented:
LGTM.

Over to @swagiaal for a 2nd opinion.

swagiaal commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
+	syncPeriod        time.Duration
+	// Local cache of failed recycle / delete operations. Map volume.Name -> status of the volume.
+	// Only PVs in Released state have an entry here.
+	releasedVolumes map[string]releasedVolumeStatus
```
Hmm.. I am not too comfortable with this approach. Wouldn't the volume state be better stored as an annotation on the PV ? You should be able to look in one place and figure out the sate of a PV. This makes it easier people other than the original author to debug volume problems. In addition if we end up using the State design pattern to clean volume states up we'll have to undo this and put it back in the volume.

markturansky replied:
See buried comment above https://github.com/kubernetes/kubernetes/pull/19365#discussion_r49107689

Using PV annotations generated many events as the PV was saved over and over. Also require more keys.  Code was smaller after switching to local state.  In the case of controller restart, all PVs start at 0, so the MaxRetries var is not guaranteed (it would be w/ PV annotation), but that doesn't seem important.

swagiaal replied:
I see... as @jsafrane has noted already though we'll pay the price of having scattered state. I wont block this though since you and Jan are aware of the trade off.

swagiaal commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 	return nil
 }
 
+// updateReleasedVolume evaluates a failed Recycle/Delete operation, updates
```
s/updateReleasedVolume/handleReleaseFailure

jsafrane replied:
Fixed

markturansky commented:
Requires rebase

swagiaal commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -125,10 +184,41 @@ func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume
 
+// updateReleasedVolume evaluates a failed Recycle/Delete operation, updates
+// internal controller state with new nr. of attempts and timestamp of the last
+// attempt. Based of nr. of failures it returns the next state of the volume
```
s/Based of nr./Based on the number/

jsafrane replied:
Fixed

jsafrane commented:
Rebased and fixed reported issues.

swagiaal commented:
One nit otherwise LGTM
@markturansky @jsafrane

jsafrane commented:
As @swagiaal 
* all variables `.*RetryCount` to `.*MaximumRetry`
* cmdline option `--pv-recycler-retry-count` to `pv-recycler-maximum-retry`

jsafrane commented:
Rebased (loosing LGTM again).

markturansky commented:
Re-adding LGTM after rebase.

jsafrane committed:
Retry recycle or delete operation on failure.

Recycle controller tries to recycle or delete a PV several times.
It stores count of failed attempts and timestamp of the last attempt in
annotations of the PV.

By default, the controller tries to recycle/delete a PV 3 times in
10 minutes interval. These values are configurable by
kube-controller-manager --pv-recycler-maximum-retry=X --pvclaimbinder-sync-period=Y
arguments.

jsafrane commented:
@markturansky, I had to regenerate docs, please review again

markturansky commented on pkg/controller/persistentvolume/persistentvolume_recycler_controller.go:
```diff
@@ -46,15 +46,33 @@ type PersistentVolumeRecycler struct {
+	// How many recycle/delete operations failed.
+	retryCount int
+	// Timestamp of the last attempt.
+	lastAttempt time.Time
```
It's interesting now that we're thinking about Conditions.  These are the kinds of fileds on a Condition (type, timestamp, etc).

markturansky commented:
re-added LGTM after doc regen.
